Added proper filters for websocket events in Mattermost connector

aboukhal opened the pull request:
# Description

Added proper filters so the bot picks up only messages that were directed to it.
Added unit tests for all the different paths and restructured the file paths for the json responses so it's easier to understand how the responses relate to each other.

Fixes #2063 


## Status
**READY**


## Type of change
- New feature (non-breaking change which adds functionality)
- Documentation (fix or adds documentation)


# How Has This Been Tested?

Running in our test environment. We removed our constraint decorator and with the new filters, it behaves exactly as before.


# Checklist:

- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation (if applicable)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

jacobtomlinson commented:
Looks like there is a Windows CI failure that needs addressing before this can be merged.

Lichtblitz commented:
One of the windows builds worked fine. I believe this is because of the flakey test suite. I assume there is an issue with not releasing the port for the mock webserver quickly enough before the next test starts. I'll look into that at some point to make tests more reliable.
Could you just try rerunning the failed tests?

jacobtomlinson commented:
Sure I've rerun them. However we might need to add `pytest-rerunfailures` to do this automatically because we don't want flaky tests on `main`.

Lichtblitz commented:
The tests have been flakey since I started work, even before the refresh. I'll check and see if I can fix the issue a bit more fine grained before going with the nuclear option.

aboukhal commented:
This is super weird. `aiohttp` raises some strange exceptions during teardown on windows+python 3.11 and windows+python 3.12 only. Timing the websocket reconnect less aggressive (1 second wait instead of adding it back to the queue without delay) has seemingly fixed it.

I've also added some debug logs to be able to analyze the elusive flakey tests related to port binding errors in the future.

I've also fixed the broken github actions in forks - those will no longer try and fail to deploy to docker hub - only to ghcr.

It's all in a separate commit for easier review.

jacobtomlinson commented on .github/workflows/deploy-docker-dev.yml:
```diff
@@ -7,7 +7,39 @@ on:
       - 'master'
 
 jobs:
+  deploy_ghcr_only:
+    if: github.repository!='opsdroid/opsdroid' # only run on forks; no need for docker hub there
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout Repos
+      uses: actions/checkout@v3
+    - name: Set up Docker Buildx
+      id: builder
+      uses: docker/setup-buildx-action@v2
+    - name: Docker meta
+      id: docker_meta
+      uses: docker/metadata-action@v3
+      with:
+        images: |
+          ghcr.io/${{ github.repository_owner }}/opsdroid
+        tags: |
+          type=raw,value=dev
+    - name: Login to GitHub Container Registry
+      uses: docker/login-action@v2
+      with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+    - name: Build and push
+      uses: docker/build-push-action@v3
+      with:
+        push: true
+        tags: ${{ steps.docker_meta.outputs.tags }}
+        platforms: linux/amd64
+        labels: ${{ steps.docker_meta.outputs.labels }}
```
Do we really need forks to publish dev containers? The intent of this action is just to run on `main` when a PR is merged so that folks can pull down the latest tip of trunk changes. I don't feel like this is something forks needs to care about.

aboukhal replied:
I use this actively to pull my work state into our test server before opening a pull request. Docker containers are much more objective for these tests, than running directly from code.

jacobtomlinson replied:
Sure, but this repo has hundreds of forks, which could result in vast amounts of storage being used. This feels like something that is concenient for you but is unnecessarily wasteful.

Perhaps I could suggest adding an extra check to the `if: github.repository!='opsdroid/opsdroid'` that looks for [repo environment variable](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#defining-configuration-variables-for-multiple-workflows). That way you can opt into this on your fork without setting it for everyone.

aboukhal replied:
Great idea, I'll look into that!

aboukhal replied:
I've added a new commit that introduces an action variable and adds documentation for it as well.

jacobtomlinson reviewed (approved):
Sweet thanks for jumping on that!
